### PR TITLE
Toolset update to 141_xp with VS2019

### PR DESCRIPTION
**Pull request by OJ**
## Description

It was well past time to get this source base updated to work a more recent toolset. So here it is!

* Solutions and projects updated so that:
    * VS 2019 builds with toolset v141_xp.
    * VS 2017 builds with toolset v141_xp.
    * VS 2013 builds with toolset v120_xp (like before).
* For development, it's recommended that you use VS 2019. If you're looking to just build new binaries, please use the updated `make.bat`.
* To facilitate the ease of building with different toolsets and platforms, I've included the jpeg lib in with the meterpreter solution now. Trust me, it's worth it! These libs get shoved in the `lib` folder.
* Output no longer writes to `output/<platformshortname>`, they all and up in `output`. @bcook-r7 and @jmartin-r7 this is going to be a key point for builds as the output folder has changed, so the gem creation stuff will need to be modified to support this.
* The new compiler resulted in a crapload more warnings, so I ended up going through them all and removing them. This included a bunch of horrible things in the python extension, and modifications to the Kiwi source as well. Most of the changes were around type casting, so hopefully nothing too shocking should be visible here.
* I had to define `DELAYIMP_INSECURE_WRITABLE_HOOKS` to get the delay load hooking working. Ultimately one day I would love to get rid of this and instead share a pointer to a "meterpreter library" of some kind. This would not only reduce the cyclomatic complexity of the application, but would also mean we can remove references to strings like 'metsrv.dll' in the source.
* PSSDK has been updated, so the sniffer extension works too. This was tested on Windows 10 and appeared to work just fine.
* I finally found and fixed a number of issues around the backcompat lib being included in x64 builds. There were two overhanging things, but the big one is that I manually modified the project files so that the reference to the lib isn't included for x64 any more. Much cleaner.
* Dependencies changed a bit. We still used OpenSSL in the Python extension for $REASONS. This has changed because updating OpenSSL was awful, and the build process sucked. Instead, we are now using LibreSSL which is smaller, and WAY easier to build. Updating this dependency will be super easy from this point on.
* The `make.bat` file has been updated to support a bunch of new stuff:
    * If you run a VS 2013 developer command prompt, you have to use v120_xp. This will be auto configured.
    * If you run a VS 2017 or VS 2019 command prompt, it will default to v141_xp. You can specify `v120_xp` as a command line parameter if you want to build against this toolset instead.

So with a developer command prompt open, you should be able to build with VS 2013/2017/2019 so long as you have the right platform toolkits installed.

image

I also enabled `/MP` on the VS projects for now. This is "multi processor builds" do make the builds work in parallel. On my VM, before these changes, a full build of all binaries took this long:

image

And the same box, after these changes:

image

It did require me to disable precompiled headers, but I think that's fine. It can't work with the Powershell extension thanks to the .NET stuff.

The binaries worked fine in my testing, and I also utilised the testing scripts that are available in the MSF repo.

Thanks to @bcook-r7 for his help in getting _ALL THE DEPS_ updated and working so this work could be done. Big thanks to @jmartin-r7 for the help getting builds sorted. Shout out to @ccondon-r7 for being great.

Hopefully this means more people can use updated toolsets to build their Meterps on Windows.

### Machine Set up
- [ ] Grab a Windows VM of some kind (10 is probably best)
- [ ] Disable Windows Defender, as it gobbles up Meterpreter binaries as they're compiled (we'll work to fix this at some point soon)
- [ ] Read the `c/meterpreter/README.md` to set up an environment correctly.
- [ ] Clone this repo and switch to this branch

### Building
- [ ] open up a Visual Studio 201X Developer command prompt and go to the `$ROOT\c\meterpeter` folder
- [ ] Update your git submodules with `git submodule update`.
- [ ] Type `make` and watch the magic happen.
- [ ] Confirm that x86 binaries build clean.
- [ ] Confirm that x64 binaries build clean.
- [ ] Copy the binaries into the `data/deps` folder in MSF, and take them for a spin.
- [ ] use all the extensions, make sure they load and commands run.
- [ ] Given that we updated the SSL in the Python extension, make sure you do some testing with SSL-related stuff in the Python extension. It appeared to work fine for me.

## Metasploit Payloads Gem

The gem build process will now need to be modified to take into account that the binaries are put in `output` and not `output/<arch>`.

## Notes

People will need to realise that these binaries are back-compat with old binaries. If people have long-term shells running, then they should not update to these binaries until those shells are done with. If you attempt to load an extension from the new batch of binaries into a shell with an old batch, then you may end up in a world of pain, suffering and torment. You don't want that. Nobody wants that.

### CODE REVIEW

**Comment by OJ** [comment]
@bcook-r7 the AppVeyor run is failing right now thanks to the update of platform toolset version.

**Comment by bcook-r7** [comment]
Great, that's what it's there for. Updating the Appveyor yaml file should allow pulling in the latest toolchain to verify. Headsup @jmartin-r7 Rapid7's internal builds will probably need a refresh as well, unless you want to switch to Appveyor for building the Windows artifacts in the first place. It does support this mode, so maybe something to consider (and perhaps one less thing for us to maintain internally!)

**Comment by jmartin-r7** [comment]
This looks promising, we can check if the source is backwards compatible with the original build process as well.  If the existing build env can build the new source already we can schedule migration at our leisure.  

Otherwise I will prioritize updating the build env generated by https://github.com/rapid7/metasploit-vagrant-builders to provide a compatible env.  Then I can look at what it takes to integrate Appveyor into the process we use internally since this all needs to be packaged for internal signing of the gem.

**Comment by jmartin-r7** [comment]
Will definitely need to adjust the build pipeline.  This PR does not like vs2013 anymore.
```
(PlatformPrepareForBuild target) -> 
13:04:34   C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.Cpp.Platform.targets(64,5): error MSB8020: The build tools for v141_xp (Platform Toolset = 'v141_xp') cannot be found. To build using the v141_xp build tools, please install v141_xp build tools.  Alternatively, you may upgrade to the current Visual Studio tools by selecting the Project menu or right-click the solution, and then selecting "Upgrade Solution...". [C:\Users\vagrant\metasploit-payloads\c\meterpreter\workspace\backcompat\backcompat.vcxproj]
```

**Comment by OJ** [comment]
Morning lads!

Thanks for getting involved @jmartin-r7! So VS 2013 will still build the source, the problem we have is with the platform toolset. VS 2013, as far as I can tell, doesn't support `141_xp` as the platform toolset, we need to go up to VS 2017 for that. So the build chain needs to be updated, and if we're going to do that we should go all the way to 2019.

My next goal is to make this all build fine with Cmake, and when we get to that point, we can easily support both VS2013 and VS2019.

But yeah, in the short term so that we can move forward, I think we're going to have to just deal with the pain of getting the build chain updated, sorry about that.

**Comment by OJ** [comment]
The updated toolset has resulted in binaries increasing in size a fair bit. The changes I made today (which I will PR soon) reduce the overall size a little, but not by much.

**Comment by OJ** [comment]
Quick look at the result of these changes shows that the file sizes are a bit different. Some changes are small enough to not worry. Others are HUGE, and are of concern. So we should go looking to see what it is that's resulted in the file sizes being so big:

image

**Comment by OJ** [comment]
Working on a solution that will allow us to build with various toolsets. Watch this space!

**Comment by OJ** [comment]
Progress ...
image
image

Am also adding a build that lets you do VS2019 with `v120_xp` (the 2013 toolset).

**Comment by OJ** [comment]
OK we have some stats. I can build clean on:
* VS 2013 with v120_xp
* VS 2019 with v120_xp
* VS 2019 with v141_xp

I'll be updating this PR with these changes as well. Interesting stats of file sizes (with a few problem children highlighted).

image

**Comment by jmartin-r7** [comment]
Thank you @OJ, new automation build test shows 2013 still expects some artifacts that are removed in this PR.  I will see if I can determine how to address this however if you have a quick fix would be appreciated.

```
17:31:36   jpeg.vcxproj -> C:\Users\vagrant\metasploit-payloads\c\meterpreter\workspace\Release\jpeg.lib
17:31:36 PostBuildEvent:
17:31:36   copy /y "C:\Users\vagrant\metasploit-payloads\c\meterpreter\workspace\Release\jpeg.lib" "C:\Users\vagrant\metasploit-payloads\c\meterpreter\source\jpeg-8\\lib\win\x86\"
17:31:36   :VCEnd
17:31:36   The system cannot find the path specified.
17:31:36           0 file(s) copied.
17:31:36 C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppCommon.targets(132,5): error MSB3073: The command "copy /y "C:\Users\vagrant\metasploit-payloads\c\meterpreter\workspace\Release\jpeg.lib" "C:\Users\vagrant\metasploit-payloads\c\meterpreter\source\jpeg-8\\lib\win\x86\" [C:\Users\vagrant\metasploit-payloads\c\meterpreter\source\jpeg-8\jpeg.vcxproj]
17:31:36 C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppCommon.targets(132,5): error MSB3073: :VCEnd" exited with code 1. [C:\Users\vagrant\metasploit-payloads\c\meterpreter\source\jpeg-8\jpeg.vcxproj]
```

**Comment by OJ** [comment]
Sorry @jmartin-r7  still WIP! Finalising the last of the changesets now. Bear with me! Want this done before breakfast :D

**Comment by OJ** [comment]
OK @jmartin-r7 we should be good to go now mate. Thanks for your patience.

**Comment by OJ** [comment]
@jmartin-r7 thanks again for your help/time in getting this stuff sorted. Please have a read of [this readme](https://github.com/rapid7/metasploit-payloads/blob/42e833c8fac5a937fa1f1d0023ed0febb80f038b/c/meterpreter/README.md) which will be part of the repo when landed. I'll be sure to add it to a wiki and make a video as some point as well. Still waiting on my new chocolatey packet to get approved, but when that's done we're good to go. I have shared the package with you via slack for the short term. Thanks!

File: c/meterpreter/README.md [hunk 1]
```diff
@@ -1,59 +1,121 @@
...
+choco install visualstudio2017community -y --package-parameters "--config Path\To\Your\.vsconfig"
+  ```
+
+Note: A copy of this file is located in this repository under `c/meterpreter/vs-config/vs2017.vsconfig`.
+
+### VS 2013
+Download and install the `Visual Studio Express 2013 for Windows Desktop` edition. It is important that you use _this exact version_. To do this with Chocolatey, run the following:
+
+```
+choco install vs2013expressdesktop5 -y
```
Comment (jmartin-r7) [inline c/meterpreter/README.md:89]:
```suggestion
choco install visualstudioexpress2013windowsdesktop -y
```

An existing approved package that meets build requirements.

Build result from `rake win_prep` on commit fda53ff
```
Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:04:44.43
```

**Comment by OJ** [comment]
Man, I can't believe I didn't find that!

Good stuff mate. Back on track now?

**Comment by OJ** [comment]
Sweet! Thank you!

**Pull request merged**